### ISSUE TITLE
Re-expand simplextrees after edge collapse

### ISSUE THREAD

**Issue by DavidLapous** [GUDHI/gudhi-devel#768]
As far as I'm aware, edge collapses are only useful for graphs and clique complexes. 
Wouldn't that make sense to re-expand by default the simplextree to its original dimension after collapsing ? That will ensure that clique complexes simplextrees preserves its homology.
I'm thinking about something like this in the `simplextree.pyx` file
```python
def collapse_edges(self, max_dimension:int=None, num:int=1, ...)->SimplexTree:
    max_dimension = self.dimension() if max_dimension is None else max_dimension
    <old code for collapse>
    self.expansion(max_dimension) # Expands back the simplextree to the original dimension.
    return self
````

**Comment by mglisse** [comment]
Thank you for this suggestion.
The main goal of edge collapses is to avoid computing the complex you would use as input in your code. That is, we only want to build triangles, tetrahedra, etc *after* doing the edge collapse. I am afraid that if we did what you suggest, some users would get confused into doing expansion *before* edge collapse, defeating the purpose.
There could be valid reasons to expand, then collapse and expand again, if only to see how many simplices you save with the collapses, but that looks like a rather niche use case, with an easy workaround (call expansion yourself afterwards), so for now I think I'd rather not do it. But I'll leave it this issue open for a while, to see if other opinions are coming.

Note that we will probably add an extra interface to edge collapses in Python, working on coo_matrix or similar.

**Comment by DavidLapous** [comment]
> I am afraid that if we did what you suggest, some users would get confused into doing expansion before edge collapse, defeating the purpose.

That's a very good point. Do you think that adding a (python) warning in that case enough to prevent users to do that ?

**Comment by mglisse** [comment]
> That's a very good point. Do you think that adding a (python) warning in that case would be enough to prevent users to do that ?

In collapse_edges, adding something like `if self.dimension() > 1: print("Warning: collapse_edges() ignores all the simplices of dimension 2 or more in this complex", file=sys.stderr)`? Yes, I think that would be ok.

**Comment by mglisse** [comment]
> As the root of the problem is an ambiguity in the method name (see [discussion](https://github.com/GUDHI/gudhi-devel/pull/1196#discussion_r2131759283_) in PR [#1196](https://github.com/GUDHI/gudhi-devel/pull/1196)), my suggestion would be to solve the problem properly by renaming the method.

Note that renaming doesn't prevent from adding a warning.

>     * `simplex_tree.collapse_as_graph(args)`: takes over the code of `simplex_tree.collapse_edges(args)`

Maybe a slightly longer name? (`collapse_edges_as_flag_graph` for a random suggestion, but I didn't really think, don't take the suggestion too strongly) "Collapse" is a rather generic term, even in Siddharth's work there was first a vertex collapse (which can also be done efficiently on flag complexes through their graph). Or did you mean it as a general "simplify" function where we don't specify what algorithm it uses? I think it may be too early for that.
I agree that adding "_as_graph" (or something similar) should make misunderstandings less likely.

>     * `simplex_tree.collapse(otherargs)`: collapses in general dimension as proposed David above

Certainly the name would have to be more specific, but that's a detail. I am still not convinced this function is a good idea, the main use case for edge-collapse is to avoid expanding before the collapse. The cases where you do want to expand and then collapse are rather niche, and the 2 persons who do want it can easily call expand explicitly.

> And the old `simplex_tree.collapse_edges(args)` would simply call `simplex_tree.collapse_as_graph(args)` with an additional `DeprecatedWarning` informing us to use the new one instead (and perhaps also the existence of the other new one).

:+1:

I think I'd be ok with this approach if the others like it.

**Comment by DavidLapous** [comment]
I agree that it's a very niche usecase (although the "weak delaunay" filtration in your pytorch optimization tuto is already a good motivation) but that's not the point IMO.
In the python spirit, things should work as expected (i.e., here : a quasi-isomorphism on flag complexes), and raise warnings if we detect bad usage. I don't want to have a broken code in the middle of a costly pipeline, in which this “optimization” is marginal. A warning is enough.

I'm fine with the renamings you proposed: 
 - current `collapse_edges` can get mapped to `collapse_edges_as_flag_graph`, with a deprecation warning
 - new `collapse` method (with possible different algorithms / strategies). 
I agree that it's a bit early as we only have 1 implemented strategy ftm, but at least this frees us a name (which will be useful later, e.g., in the multipersistence setting).
Would something like this be ok for you  @mglisse @VincentRouvreau @hschreiber @MathieuCarriere  ?
```python
available_strategies = Literal["edge_collapse"]
def collapse(self, expansion_dim:int=None, strategy:available_strategies="edge_collapse", ...)->SimplexTree:
    """
    some doc
    """
    if strategy == "edge_collapse":
      if expansion_dim is None and self.dimension() > 1:
        from warnings import warn
        warn(f"(Unnecessary computation) {strategy=} assumes this complex is a flag complex and ignore simplices of dimension>1.") 
      max_dimension = self.dimension() if expansion_dim is None else expansion_dim
      collapsed_complex = self.collapse_edges_as_graph(...)
      collapsed_complex.expansion(max_dimension)
      return collapsed_complex
    raise ValueError(f"Unavailable strategy {strategy}. Available ones are {available_strategies}.")
```
Something unrelated : I think this should be a const method returning a new collapsed cplx, as we're already doing a copy (hence the `collapsed_complex` instead of `self`).
